### PR TITLE
Feature/consolidated tools improvement

### PR DESCRIPTION
**Pull request by shashax42**
## What does this PR do?
Fixes consolidated tool discovery default behavior and aligns unit test command registration with production.

- Treats `ServiceStartOptions.ReadOnly == null` as `false` during command filtering, so the default (unset) behavior does not unintentionally filter down to read-only commands.
- Updates the unit test [CommandFactoryHelpers](cci:2://file:///c:/Users/PC/OneDrive/%EB%B0%94%ED%83%95%20%ED%99%94%EB%A9%B4/MS/mcp/core/Azure.Mcp.Core/tests/Azure.Mcp.Core.UnitTests/Areas/Server/CommandFactoryHelpers.cs:71:0-240:1) registration so the same tool areas are registered as in production, preventing gaps in consolidated tool validation coverage.
- Adds unit tests covering consolidated tool mapping completeness, expected command counts with all areas, and metadata equality edge cases.

**Notes for reviewers**
- This is a narrow behavior fix: only the `ReadOnly` default handling changes (null → false).
- All unit tests pass locally: `dotnet test core/Azure.Mcp.Core/tests/Azure.Mcp.Core.UnitTests -c Debug` (852 tests).

## Relevant links
- Contribution guidelines: https://github.com/microsoft/mcp/blob/main/CONTRIBUTING.md
- Consolidated tools definition: https://github.com/microsoft/mcp/blob/main/core/Microsoft.Mcp.Core/src/Areas/Server/Resources/consolidated-tools.json
- Consolidated tool discovery strategy: https://github.com/microsoft/mcp/blob/main/core/Microsoft.Mcp.Core/src/Areas/Server/Commands/Discovery/ConsolidatedToolDiscoveryStrategy.cs
- Commit cleanup guide: https://github.com/Azure/azure-powershell/blob/main/documentation/development-docs/cleaning-up-commits.md
- ToolDescriptionEvaluator quickstart (reference): https://github.com/microsoft/mcp/blob/main/eng/tools/ToolDescriptionEvaluator/Quickstart.md

## GitHub issue number?
#1759 

## Pre-merge Checklist
- [x] Required for All PRs
  - [x] **Read [contribution guidelines](https://github.com/microsoft/mcp/blob/main/CONTRIBUTING.md)**
  - [x] PR title clearly describes the change
  - [x] Commit history is clean with descriptive messages ([cleanup guide](https://github.com/Azure/azure-powershell/blob/main/documentation/development-docs/cleaning-up-commits.md))
  - [x] Added comprehensive tests for new/modified functionality
  - [ ] Updated `servers/Azure.Mcp.Server/CHANGELOG.md` and/or `servers/Fabric.Mcp.Server/CHANGELOG.md` for product changes (`features, bug fixes, UI/UX, updated dependencies`)  
        Not applicable — no product docs/dependency updates, and no tool surface changes beyond correcting the default option interpretation.

- [ ] For MCP tool changes:
  - [ ] **One tool per PR**: This PR adds or modifies only one MCP tool for faster review cycles  
        Not applicable — this PR changes consolidated discovery behavior and test infrastructure, not an individual tool implementation.
  - [ ] Updated `servers/Azure.Mcp.Server/README.md` and/or `servers/Fabric.Mcp.Server/README.md` documentation  
        Not applicable.
  - [ ] Validate README.md changes using script at `eng/scripts/Process-PackageReadMe.ps1`. See [Package README](https://github.com/microsoft/mcp/blob/main/CONTRIBUTING.md#package-readme)  
        Not applicable.
  - [ ] Updated command list in [/servers/Azure.Mcp.Server/docs/azmcp-commands.md](cci:7://file:///c:/Users/PC/OneDrive/%EB%B0%94%ED%83%95%20%ED%99%94%EB%A9%B4/MS/mcp/servers/Azure.Mcp.Server/docs/azmcp-commands.md:0:0-0:0) and/or `/docs/fabric-commands.md`  
        Not applicable.
  - [ ] Run `.\eng\scripts\Update-AzCommandsMetadata.ps1` to update tool metadata in azmcp-commands.md (required for CI)  
        Not applicable.
  - [ ] For new or modified tool descriptions, ran [`ToolDescriptionEvaluator`](https://github.com/microsoft/mcp/blob/main/eng/tools/ToolDescriptionEvaluator/Quickstart.md) and obtained a score of `0.4` or more and a top 3 ranking for all related test prompts  
        Not applicable.
  - [ ] For tools with new names, including new tools or renamed tools, update [[consolidated-tools.json](cci:7://file:///c:/Users/PC/OneDrive/%EB%B0%94%ED%83%95%20%ED%99%94%EB%A9%B4/MS/mcp/core/Microsoft.Mcp.Core/src/Areas/Server/Resources/consolidated-tools.json:0:0-0:0)](https://github.com/microsoft/mcp/blob/main/core/Microsoft.Mcp.Core/src/Areas/Server/Resources/consolidated-tools.json)  
        Not applicable.
  - [ ] For new tools associated with Azure services or publicly available tools/APIs/products, add URL to documentation in the PR description  
        Not applicable.

- [ ] Extra steps for **Azure MCP Server** tool changes:
  - [ ] Updated test prompts in `/servers/Azure.Mcp.Server/docs/e2eTestPrompts.md`  
        Not applicable.
  - [x] 👉 For Community (non-Microsoft team member) PRs:
    - [x] **Security review**: Reviewed code for security vulnerabilities, malicious code, or suspicious activities before running tests (`crypto mining, spam, data exfiltration, etc.`)
    - [ ] **Manual tests run**: added comment `/azp run mcp - pullrequest - live` to run *Live Test Pipeline*  
          Not run — unit-test scoped change. Local unit tests were run (852 tests).

### CODE REVIEW

**Comment by vcolin7** [comment]
Hey @fanyang-mono! It looks like this is a community contribution that could use your expertise for a review. Mind taking a look when you get a moment? Thanks!

**Comment by anuchandy** [comment]
Adding 'Do Not Merge' to filter it out from the 2.0 cutoff review queue. Please remove the label when it's close to ready and once Fang reviewed it.

File: core/Microsoft.Mcp.Core/src/Areas/Server/Commands/Discovery/ConsolidatedToolDiscoveryStrategy.cs [hunk 1]
```diff
@@ -78,7 +78,7 @@ public ICommandFactory CreateConsolidatedCommandFactory()
 
 #if DEBUG
             // In debug mode, validate that all tools in MappedToolList found a match when conditions are met
-            if (_options.Value.ReadOnly == false && (_options.Value.Namespace == null || _options.Value.Namespace.Length == 0))
+            if ((_options.Value.ReadOnly ?? false) == false && (_options.Value.Namespace == null || _options.Value.Namespace.Length == 0))
```
Comment (fanyang-mono) [inline core/Microsoft.Mcp.Core/src/Areas/Server/Commands/Discovery/ConsolidatedToolDiscoveryStrategy.cs:85]: @shashax42 This is different than the original design. Please explain the intent of this code change to be considered.

**Review by fanyang-mono — CHANGES_REQUESTED** [review]

File: core/Azure.Mcp.Core/tests/Azure.Mcp.Core.UnitTests/Areas/Server/CommandFactoryHelpers.cs [hunk 2]
```diff
@@ -82,14 +112,19 @@ public static ICommandFactory CreateCommandFactory(IServiceProvider? serviceProv
...
             new QuotaSetup(),
             new RedisSetup(),
             new ResourceHealthSetup(),
             new SearchSetup(),
             new ServiceBusSetup(),
+            new SignalRSetup(),
+            new SpeechSetup(),
             new SqlSetup(),
             new StorageSetup(),
+            new StorageSyncSetup(),
```
Comment (jongio) [inline core/Azure.Mcp.Core/tests/Azure.Mcp.Core.UnitTests/Areas/Server/CommandFactoryHelpers.cs]: **[MEDIUM]** Tests: 5 tool areas from production `Program.cs` are still missing here: `ContainerAppsSetup`, `DeviceRegistrySetup`, `FunctionsSetup`, `ServiceFabricSetup`, `WellArchitectedFrameworkSetup`.

The `MapsAllRegisteredCommands` test computes its expected count from this same list, so it can't detect drift from production. Adding the missing areas would close that gap.

**Review by jongio — COMMENTED** [review]
Fixes a genuine bug: `ServiceStartOptions.ReadOnly` is `bool?` defaulting to `null`, and `null == false` is `false` in C# - so the default accidentally restricted to read-only commands. The null-coalescing fix is correct and consistent across both usages.

Issues to address:
- CommandFactoryHelpers.cs:127 - test helper still misses 5 production areas (ContainerApps, DeviceRegistry, Functions, ServiceFabric, WellArchitectedFramework)

Branch needs rebase - main has diverged (Foundry to FoundryExtensions rename, service mock registration changes).

File: core/Microsoft.Mcp.Core/src/Areas/Server/Commands/Discovery/ConsolidatedToolDiscoveryStrategy.cs (hunk 1 above)
Comment (shashax42) [reply core/Microsoft.Mcp.Core/src/Areas/Server/Commands/Discovery/ConsolidatedToolDiscoveryStrategy.cs:85]: @fanyang-mono Thanks for the question.
(Sorry for the late reply. I missed the notification while I was caught up with other things!)

This change is not intended to alter the original design, but to make consolidated discovery consistent with the existing behavior in the rest of the server-start pipeline.

`ServiceStartOptions.ReadOnly` is a nullable bool that defaults to `null`. In other parts of the codebase, such as `CommandGroupDiscoveryStrategy`, `ServiceCollectionExtensions`, and the existing null/default read-only tests, a `null` value is already treated as `false`.

Consolidated discovery was the outlier here: it used `== false`, which excludes the `null` case in C#. As a result, the default unset behavior could unintentionally filter to only read-only commands.

Using `ReadOnly ?? false` brings consolidated discovery in line with the established behavior, restores the intended default semantics for the unset case, and preserves explicit `--read-only` behavior unchanged.

I’ll also add regression test coverage for the null/default case to make that behavior explicit.

File: core/Azure.Mcp.Core/tests/Azure.Mcp.Core.UnitTests/Areas/Server/CommandFactoryHelpers.cs (hunk 2 above)
Comment (shashax42) [reply core/Azure.Mcp.Core/tests/Azure.Mcp.Core.UnitTests/Areas/Server/CommandFactoryHelpers.cs]: @jongio Thanks for the review!

I rebased the branch onto the latest current main and updated the test helper registrations to match the current production area list. The helper now matches the production tool area registrations on current main, and I refreshed the consolidated discovery coverage accordingly.

I also added explicit regression coverage for the null/default ReadOnly case so the intended default behavior is now covered directly.


Also, I really appreciate everyone who took the time to review this so thoroughly , it means a lot. Thank you!

**Review by jongio — COMMENTED** [review]
Addresses my previous feedback. The rebase onto latest main looks clean, and the missing tool areas I flagged (ContainerApps, DeviceRegistry, Functions, ServiceFabric, WellArchitectedFramework) are all included now - along with the other areas that were also missing. The test helper's area list now matches production exactly.

The `ReadOnly` null-coalescing fix is correct. `ServiceStartOptions.ReadOnly` is `bool?` defaulting to `null`, and the old `== false` check excluded `null` (since `null == false` is `false` in C#). Five other call sites in the codebase already use `?? false` for this field (`ServiceCollectionExtensions`, `CommandGroupDiscoveryStrategy`, `NamespaceToolLoader`, `ServerToolLoader`, `McpRuntime`), so this just brings `ConsolidatedToolDiscoveryStrategy` in line.

The `WithNullReadOnlyOption_DefaultsToFalse` test is a good regression test for this specific bug - it directly validates that `null` and `false` produce identical command sets.

One suggestion: this behavioral fix (default `ReadOnly` silently filtering to read-only commands in consolidated mode) seems worth a CHANGELOG entry under bug fixes, since users hitting this bug would see fewer tools than expected with no obvious cause.

Note: the PR still has the "Do Not Merge" label from the 2.0 cutoff triage. That'll need removing when @fanyang-mono is satisfied with the author's response to the original design question.

**Review by jongio — COMMENTED** [review]
My previous feedback is addressed. The duplicated area list is extracted to `CreateAreaSetups()`, the unused import is removed, and the brittle count test is replaced with reference-equality mapping validation. Clean changes.

The branch needs another rebase though. Since the last rebase (April 28), main has diverged:

- `AzureBackupSetup` and `AzureTerraformSetup`: these projects no longer exist on main. The test helper references them but they'll cause build failures on merge.
- `ContainerAppsSetup`: the project directory exists but `src/` is empty (no Setup class).
- `MonitorInstrumentationSetup`: added to main (Program.cs line 128) but missing from the test helper.

The `MapsAllRegisteredCommands` test won't catch this drift because it validates against what the test helper registers, not against production Program.cs. After rebase, the area list will need updating again.

Note: @fanyang-mono's changes-requested review from March is still pending. The author's April 28 response explaining the `?? false` rationale looks correct (5 other call sites use the same pattern).

**Review by jongio — COMMENTED** [review]
Incremental check (rebase + 1 new commit since my last review).

The new area registrations in `CreateAreaSetups()` are a solid step forward. 19 missing tool areas added, which will make the `MapsAllRegisteredCommands` test much more meaningful.

Three areas in the test helper still don't exist on `main` (no source files, not registered in `Program.cs`):
- `AzureBackupSetup`
- `AzureTerraformSetup`
- `ContainerAppsSetup`

These will cause build failures once you rebase onto current `main`. Drop them from `CreateAreaSetups()`.

One production area is still missing from the test helper:
- `MonitorInstrumentationSetup` (registered in `Program.cs`, has source files under `tools/Azure.Mcp.Tools.MonitorInstrumentation/`)

Add it so the mapping test actually catches future drift.

Previous blockers remain: merge conflicts need resolving, fanyang-mono's review is still outstanding, and the "Do Not Merge" label is still applied.

**Comment by shashax42** [comment]
@jongio 
Currently, based on upstream/main, AzureBackupSetup, AzureTerraformSetup, and ContainerAppsSetup are all registered in Program.cs and their source files exist. As for MonitorInstrumentationSetup, it doesn't have a separate Setup class but is included within Monitor. Could you please check if your comments were based on changes that haven't been merged yet?

File: core/Microsoft.Mcp.Core/src/Areas/Server/Commands/Discovery/ConsolidatedToolDiscoveryStrategy.cs [hunk 3]
```diff
@@ -82,7 +82,7 @@ public ICommandFactory CreateConsolidatedCommandFactory()
...
```
Comment (wbreza) [inline core/Microsoft.Mcp.Core/src/Areas/Server/Commands/Discovery/ConsolidatedToolDiscoveryStrategy.cs:85]: [F-001] **LOW — Cover the debug-only null guard**

The null/default regression test exercises `FilterCommands`, but it cannot detect a regression in this separately changed `#if DEBUG` validation guard because the real definition provider has no missing mappings. Reverting this guard alone would leave every added test green. Add a DEBUG test whose definition provider returns one missing mapped command under `ReadOnly = null`, then assert `InvalidOperationException`.

File: core/Microsoft.Mcp.Core/src/Areas/Server/Commands/Discovery/ConsolidatedToolDiscoveryStrategy.cs [hunk 4]
```diff
@@ -175,7 +175,7 @@ private Dictionary<string, IBaseCommand> FilterCommands(IReadOnlyDictionary<stri
                 var serviceArea = _commandFactory.GetServiceArea(kvp.Key);
                 return serviceArea == null || !IgnoredCommandGroups.Contains(serviceArea, StringComparer.OrdinalIgnoreCase);
             })
-            .Where(kvp => _options.Value.ReadOnly == false || kvp.Value.Metadata.ReadOnly == true)
+            .Where(kvp => (_options.Value.ReadOnly ?? false) == false || kvp.Value.Metadata.ReadOnly == true)
```
Comment (wbreza) [inline core/Microsoft.Mcp.Core/src/Areas/Server/Commands/Discovery/ConsolidatedToolDiscoveryStrategy.cs:178]: [O-002] 💡 Improvement Opportunity (optional — not required to merge)

**Benefit:** Both nullable Boolean checks state their null-as-false intent directly.
**Effort:** Small.
**Optionality:** The current expressions are correct.
**Sketch:** Use `_options.Value.ReadOnly is not true` at lines 85 and 178.

File: core/Azure.Mcp.Core/tests/Azure.Mcp.Core.Tests/Areas/Server/Commands/Discovery/ConsolidatedToolDiscoveryStrategyTests.cs [hunk 5]
```diff
@@ -47,77 +47,162 @@ private static ConsolidatedToolDiscoveryStrategy CreateStrategy(
...
+            Secret = false,
+            LocalRequired = false
+        };
+        Assert.True(ConsolidatedToolDiscoveryStrategy.AreMetadataEqual(metadata1, metadata2));
+    }
+
+    [Fact]
+    public void AreMetadataEqual_DifferentValues_ReturnsFalse()
+    {
+        var metadata1 = new ToolMetadata { Destructive = false, ReadOnly = true };
```
Comment (wbreza) [inline core/Azure.Mcp.Core/tests/Azure.Mcp.Core.Tests/Areas/Server/Commands/Discovery/ConsolidatedToolDiscoveryStrategyTests.cs:204]: [O-003] 💡 Improvement Opportunity (optional — not required to merge)

**Benefit:** A data-driven test that flips each `ToolMetadata` property catches future omissions from `AreMetadataEqual`.
**Effort:** Small.
**Optionality:** The current implementation and tests are correct.
**Sketch:** Convert the differing-values case to a theory covering all six compared properties.

**Review by wbreza — APPROVED** [review]
### What stood out
- **Consistent default behavior** — An unset `ReadOnly` option now behaves like `false` across consolidated discovery, matching the documented contract and the rest of server startup.
- **Stronger mapping coverage** — Shared test registrations and reference-equality assertions protect the consolidated command set from omissions.

### Top issues
- [F-001] Add regression coverage for the debug-only null guard — `core/Microsoft.Mcp.Core/src/Areas/Server/Commands/Discovery/ConsolidatedToolDiscoveryStrategy.cs:85`

### Improvement opportunities
- [O-002] Simplify nullable Boolean checks _(Optional)_ — `core/Microsoft.Mcp.Core/src/Areas/Server/Commands/Discovery/ConsolidatedToolDiscoveryStrategy.cs:178`
- [O-003] Exercise every metadata comparison _(Optional)_ — `core/Azure.Mcp.Core/tests/Azure.Mcp.Core.Tests/Areas/Server/Commands/Discovery/ConsolidatedToolDiscoveryStrategyTests.cs:204`
